contracts/policy-agents/PolicyOfferExecutor.sol

```diff
@@ -0,0 +1,87 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../interfaces/IClaimsManager.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+// This contract needs to have the admin role for the ClaimsManager contract.
+// It can only call updatePolicy() with the permission of the claimant, which
+// effectively makes this contract a policy agent.
+contract PolicyOfferExecutor {
+    using ECDSA for bytes32;
+    using SafeERC20 for IERC20;
+
+    address public immutable claimsManager;
+    address public immutable offerSigner;
+    address public immutable token;
+    address public immutable beneficiary;
+    mapping(bytes32 => bool) public offerWithIdIsExecuted;
+
+    constructor(
+        address _claimsManager,
+        address _offerSigner,
+        address _token,
+        address _beneficiary
+    ) {
+        claimsManager = _claimsManager;
+        offerSigner = _offerSigner;
+        token = _token;
+        beneficiary = _beneficiary;
+    }
+
+    function executeOffer(
+        bytes32 offerId,
+        bytes[] calldata policyData,
+        uint256 offerAmount,
+        uint256 offerExpiration,
+        bytes calldata offerSignature
+    ) external returns (bytes32[] memory policyHashes) {
+        require(
+            offerId ==
+                keccak256(abi.encode(policyData, offerAmount, offerExpiration)),
+            "Invalid offer ID"
+        );
```

Comment:
Is it better to calculate this hash separately instead of just adding the `offerId` to the signature and get it validated below in lines 47-60?

Reply:
Also shouldn't this be called `offerHash` (since we used `policyHash` instead of `policyId` I mean)?

Reply:
> Also shouldn't this be called offerHash (since we used policyHash instead of policyId I mean)?

Yeah I'm aware. My reasoning was that this is a user-facing name (when a multisig tx is created you're supposed to copy-paste the order ID to the Market to see what it's about) and people may not be able to figure out that `orderHash` is what they should use to identify their order.
I'm also aware of the inconsistency, but my solution there was to rename all `policyHash` and `claimHash` to `policyId` and `claimId` https://github.com/api3dao/claims-manager/issues/93